src/rust/ide/src/model/synchronized/module.rs

```diff
@@ -0,0 +1,381 @@
+            Err(err)        => {
+                error!(self.logger,"Error during sending text change to Language Server: {err}");
+                LanguageServerContent::Desynchronized(old_content)
+            }
+        }
+    }
+
+    /// Handle received notification. Returns the new content summery of Language Server state.
+    async fn handle_notification
+    (&self, content:&LanguageServerContent, notification:Notification)
+    -> FallibleResult<ParsedContentSummary> {
+        match content {
+            LanguageServerContent::Desynchronized(summary) => self.full_invalidation(summary).await,
+            LanguageServerContent::Synchronized(summary)   => match notification {
+                Notification::Invalidate => self.full_invalidation(&summary.summary).await,
+                Notification::CodeChanged{change,replaced_location} =>
+                    self.notify_language_server(&summary.summary, |content| {
+                        let code_change = TextEdit {
+                            range : replaced_location.into(),
+                            text  : change.inserted,
+                        };
+                        let id_map_change = TextEdit {
+                            range : summary.id_map.clone().into(),
+                            text  : content.id_map_slice().to_string(),
+                        };
+                        vec![code_change,id_map_change]
+                    }).await,
+                Notification::MetadataChanged =>
+                    self.notify_language_server(&summary.summary, |content| vec![TextEdit {
+                        range : summary.metadata.clone().into(),
+                        text  : content.metadata_slice().to_string(),
+                    }]).await,
+            },
+        }
+    }
+
+    /// Send update to Language Server with the entire file content. Returns the new content summary
+    /// of Language Server state.
+    async fn full_invalidation
+    (&self, ls_content:&ContentSummary) -> FallibleResult<ParsedContentSummary> {
+        let range = TextLocation::at_document_begin()..ls_content.end_of_file;
+        self.notify_language_server(ls_content,|content| vec![TextEdit {
+            range : range.into(),
+            text  : content.string
+        }]).await
+    }
+
+    /// This is a helper function with all common logic regarding sending the update to
+    /// Language Server. Returns the new summary of Language Server state.
+    async fn notify_language_server
+    ( &self
+    , ls_content        : &ContentSummary
+    , edits_constructor : impl FnOnce(SerializedSourceFile) -> Vec<TextEdit>
+    ) -> FallibleResult<ParsedContentSummary> {
+        let content = self.model.serialized_content()?;
+        let summary = ParsedContentSummary::from_source(&content);
+        let edit    = language_server::types::FileEdit {
+            path        : self.path.file_path().clone(),
+            edits       : edits_constructor(content),
```

Comment:
What are the rules for applying multiple edits in the same batch?
Do subsequent edits are described in positions relative to the original contents or relative to the result of previous edit? 

Have we tested this / do we comply with whatever engine is doing here?

Reply:
tested it and apparently subsequent edits' positions should be those "affected" by previous edits. I fixed that.

Reply:
I was sure this was documented! Apparently it's not, I'll add the doc for it. You are correct – multiple edits in one message behave the same as multiple messages with one edit each.